include/mlir/IR/StandardTypes.h

```diff
@@ -422,6 +424,38 @@ class MemRefType
   using Base::getImpl;
 };
 
+/// Unranked MemRef type represent multi-dimensional MemRefs that
+/// have an unknown rank.
```

Comment:
I don't like the fact that this doesn't also inherit from MemRef. If you are going to add this, you should mirror what we already do for TensorType.

Reply:
All of the other work to add a 'RankedMemRefType' should come before this, otherwise merging the two later into one class hierarchy is going to be extremely difficult.

Reply:
What do you suggest regarding class names ? Renaming `MemRefType` to `RankedMemRefType` will touch a lot of file and docs. `BaseMemRefType`, `MemRefType` and `UnrankedMemRefType` ?

Reply:
MemRefType, RankedMemRefType, UnRankedMemRefType to match what we do for TensorTypes. Whatever you do should match the naming of the TensorType hierarchy.

Reply:
This is a necessary part of changing/evolving a codebase to ensure that everything remains clean and consistent.

Reply:
#268 renames MemRefType to RankedMemRefType. I will update this PR once #268 is in to add the class hierarchy.

Reply:
I'll try to iterate quickly on it, please also mark it "- NFC" and rebase so I can start our internal tests.

Reply:
Sorry, I'm late to this. But this is what I posted at #268. 
To keep names and lines short and because the ranked MemRefType is far more widespread in use across the codebase, would it make sense to just continue to use MemRefType for a ranked memref and use RankErasedMemRefType or UnrankedMemRefType for an unranked one? Note that the use case for an unranked tensor type (eg. shape inference, etc. in more abstract levels of the IR) are quite different from that of an unranked memref type(?) and so it isn't strictly needed to maintain naming consistency across the two. (IIUC, unranked memref types are just for interfacing.) If we do want to maintain consistency with TensorType, it looks more compact to do RankedTensorType -> TensorType, and go with BaseMemrefType/BaseTensorType or MemRefBaseType/TensorBaseType for the base class?

Reply:
I am actually +1 on not changing the name like @bondhugula suggests.
I agree with keeping the same hierarchy as for Tensors but using the naems @bondhugula proposes.
Using the same naming system as for Tensor goes against the "remains clean and consistent" objective: unranked tensor and rankerasedmemrefs are fundamentally different concepts in their current and (foreseeable) future usage.

Reply:
I’m against this proposal for several reasons, but I won’t be able to fully articulate them given limited availability this week. For now:
a) Keeping names short and trying to avoid churn(updating API/Docs)  are *not* valid reasons to not do something IMO. We should always be striving to do the “right” thing, and these feels like corner cutting justifications to avoid the need to do more work.
b) The suggesting naming doesnt make sense to me and seems very likely to breed confusion. “Why is UnrankedTensorType not a TensorType”.

tldr: I’d rather us do the right thing instead of the easiest/most compact.

Reply:
> I’m against this proposal for several reasons, but I won’t be able to fully articulate them given limited availability this week. For now:
> a) Keeping names short and trying to avoid churn(updating API/Docs) are _not_ valid reasons to do something IMO. We should always be striving to do the “right” thing, and these feels like corner cutting justifications to avoid the need to do more work.
> b) The suggesting naming doesnt make sense to me and seems very likely to breed confusion. “Why is UnrankedTensorType not a TensorType”.

FWIW, the line of reasoning is as follows. A memref is always ranked (there had been no such thing as a rankless memref because dimensionality is a core part of the memref). As such, using MemRefType to always refer to the known rank case isn't out of line I think. An AbstractMemRefType or BaseMemRefType (tend to prefer the former) could serve as the common base class for both the known rank and the unknown rank abstractions. I think it's fine to say that an UnrankedMemRefType is an AbstractMemRefType but not a MemRefType (but the adjective 'unranked' is at odds here if viewed in one way, and so RankErasedMemRef might appear better to avoid confusion). In fact, the unranked memref abstraction shouldn't have a shape (because it doesn't even have a rank) -- same is the case with an UnrankedTensorType, and so it's odd that UnrankedTensorType is currently deriving from ShapedType. Most of the doc comments on mlir::ShapedType methods are actually prefixed with "If this is a ranked type, ..." and they abort for UnrankedTensorType's. For the UnrankedMemRefType, if possible, we shouldn't derive from a ShapedType.

Reply:
Today we have the hierarchy:
```
ShapedType --- TensorType --- UnrankedTensorType
                 |----------- RankedTensorType
```

Assuming the hierarchy:
```
ShapedType --- MemRefType --- UnrankedMemRefType
                 |----------- RankedMemRefType
```

I claim there will not be a use case in the near/mid-term future that will want to do  .dyn_cast<MemRefType>().

I would even go as far as saying maybe the proper type hierarchy should be:
```
ShapedType --- TensorType
UnshapedType --- UnrankedTensorType
```

Uday's remark suggests that for existing use cases of Tensor the current class hierarchy imposes pure boilerplate on the users.
There may be other concrete use cases I am not aware of so I am not 100% sure here.

However for the MemRefType, the following class hierarchy is much more in tune with the use cases we will have:
```
ShapedType --- MemRefType
UnshapedType --- RankErasedMemRefType
```

I do not see the value brought by trying to force MemRefType in a class hierarchy that is unnatural for the foreseeable use cases.

Reply:
To be more mathematically precise, I would say that `UnshapedType` is **the** mathematically unique empty shape (a singleton type).

Reply:
This was discussed and the results of the discussion were posted to the main thread.